File: median_of_two_sorted_arrays.py
# Time Complexity: O(log(m+n)) where m and n is length of lists
#  Space Complexity:  O(1).
#  Did this code successfully run on Leetcode : Yes
#  Any problem you faced while coding this : No



class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        len_nums1 = len(nums1)
        len_nums2 = len(nums2)
        # to make the smaller length array at the beginning. # Binary Search on smaller array.
        if(len_nums2 < len_nums1):
            return self.findMedianSortedArrays(nums2, nums1)
        low = 0
        high = len_nums1

        while (low <= high):
            part_x = (low + high) // 2
            part_y = (len_nums1+ len_nums2) // 2 - part_x

            l1 = float("-inf") if part_x == 0 else nums1[part_x-1]
            l2 = float("-inf") if part_y == 0 else nums2[part_y-1]
            r1 = float("inf") if part_x == len_nums1 else nums1[part_x]
            r2 = float("inf") if part_y == len_nums2 else nums2[part_y]

            if(l1 <= r2 and l2 <= r1):
                if((len_nums1 + len_nums2) % 2 == 0 ):
                    return (max(l1,l2) + min(r1,r2)) / 2 
                return min(r1,r2)
            
            if (l2 > r1):
                low = part_x + 1
                
            elif l1 > r2:
                high = part_x-1
        
        return 0

File: Median of Two Sorted Arrays.cpp
//Time Complexity- O(log(min(m,n))
//Space Complexity- O(1)

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        
        if(nums1.size()==0 && nums2.size()==0){
            return 0.00;
        }
        int n1=nums1.size();
        int n2=nums2.size();
        if(n1>n2){
            return findMedianSortedArrays(nums2,nums1);
        }
        
        int low=0;
        int high=n1;
        while(low<=high){
            int px=low+(high-low)/2;
            int py=(n1+n2)/2-px;
            double l1=(px==0)?INT_MIN:nums1[px-1];
            double l2=(py==0)?INT_MIN:nums2[py-1];
            double r1=(px==n1)?INT_MAX:nums1[px];
            double r2=(py==n2)?INT_MAX:nums2[py];
            
            if(l1<=r2 && l2<=r1){
                if((n1+n2)%2==0){
                    return (max(l1,l2)+min(r1,r2))/2;
                }
                else{
                    return min(r1,r2);
                }
            }
            else if(l1>r2){
                high=px-1;
            }
            else{
                low=px+1;
            }
        }
        return -1;
    }
};